File: algorithm/543.cpp
#include <cstddef>
#include <algorithm>
using namespace std;

struct TreeNode
{
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x): val(x), left(NULL), right(NULL) {};
};

class Solution
{
public:
        Solution():maxPathLen(0) {}
        int diameterOfBinaryTree(TreeNode* root);

private:
        int maxPathLen;
        int postorder(TreeNode* root);
};

int Solution::diameterOfBinaryTree(TreeNode* root)
{
        if(root==NULL || root->left==NULL && root->right==NULL) return 0;
    
        int tmp = postorder(root);
        return maxPathLen;

}

int Solution::postorder(TreeNode* root)
{
        if(root==NULL) return 0;

        int lheight = postorder(root->left); 
        int rheight = postorder(root->right); 
        int pathlen = lheight + rheight;
        if(pathlen>maxPathLen) maxPathLen = pathlen;

        return max(lheight, rheight)+1;

}

int main()
{
	return 0;
}

File: algorithm/go/firstround/543.go
package firstround

//func diameterOfBinaryTree(root *TreeNode) int {
//	if root == nil {
//		return 0
//	}
//	res := 0
//	traverse(root, &res)
//	return res
//}
//
//func traverse(node *TreeNode, res *int) int {
//	if node == nil {
//		return 0
//	}
//
//	leftMaxEdgeNum, rightMaxEdgeNum := 0, 0
//	if node.Left != nil {
//		leftMaxEdgeNum = traverse(node.Left, res) + 1
//	}
//
//	if node.Right != nil {
//		rightMaxEdgeNum = traverse(node.Right, res) + 1
//	}
//
//	*res = max(*res, leftMaxEdgeNum+rightMaxEdgeNum)
//
//	return max(leftMaxEdgeNum, rightMaxEdgeNum)
//}
//
//func max(a, b int) int {
//	if a < b {
//		return b
//	}
//	return a
//}

func diameterOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}
	res := 0
	traverse(root, &res)
	return res
}

func traverse(node *TreeNode, res *int) int {
	if node == nil {
		return 0
	}

	leftMaxDepth := traverse(node.Left, res)
	rightMaxDepth := traverse(node.Right, res)

	*res = max(*res, leftMaxDepth+rightMaxDepth)

	return max(leftMaxDepth, rightMaxDepth) + 1
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}
